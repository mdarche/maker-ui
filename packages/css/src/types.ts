import type { Interpolation } from '@emotion/react'

export type Breakpoints = (string | number)[]
export type StyleObject = object | Interpolation<any>

/**
 * Support for the breakpoints prop and responsive CSS scales.
 *
 * @param breakpoints - An array of breakpoints that determines the css prop media queries
 * @param css - A css object that can support array values for all nested properties
 *
 * @remarks
 * Used by all Maker UI components, primitives, and custom jsx pragma.
 *
 */
export interface MakerProps {
  css?: StyleObject
  breakpoints?: Breakpoints
}

/**
 * A string, number, or array of either lets you stack values for media queries.
 *
 * @example A simple example:
 * <div css={{ height: [50, 100]}} />
 */
export type ResponsiveScale = string | number | (string | number)[]

declare module 'react' {
  interface Attributes extends MakerProps {}
}

/**
 * Direct export from EmotionJS
 */
export { Interpolation }
