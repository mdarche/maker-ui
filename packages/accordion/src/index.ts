export { Accordion } from './Accordion'
export { AccordionPanel } from './AccordionPanel'
