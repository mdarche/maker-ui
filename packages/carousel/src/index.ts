export { Carousel, CarouselProps } from './Carousel'
