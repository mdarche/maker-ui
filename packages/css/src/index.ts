export { Global } from './Global'
export type { GlobalProps } from './Global'
export { jsx } from './jsx-runtime'

export type {
  Breakpoints,
  StyleObject,
  MakerProps,
  ResponsiveScale,
  ResponsiveString,
  Interpolation,
} from './types'
