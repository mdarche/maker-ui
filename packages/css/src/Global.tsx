import * as React from 'react'
import { Global as EmotionGlobal } from '@emotion/react'

import { formatCSS } from './css'
import type { Interpolation } from './types'

export interface GlobalProps {
  styles: Interpolation<any> | object
  breakpoints?: (string | number)[]
}

/**
 * A global style declaration that supports responsive arrays and the `breakpoints` prop
 *
 * @param styles - a CSS style object
 * @param breakpoints - an array of breakpoints
 *
 * @link https://maker-ui.com/docs/elements/global
 */
export const Global = ({ styles, breakpoints }: GlobalProps) => {
  return (
    <EmotionGlobal
      styles={
        formatCSS(
          styles as Interpolation<any>,
          breakpoints
        ) as Interpolation<any>
      }
    />
  )
}
