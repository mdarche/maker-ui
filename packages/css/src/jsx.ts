import { createElement } from 'react'
import { jsx as emotionJsx, Interpolation } from '@emotion/react'

import { formatCSS } from './css'

/**
 * Scans JSX props for `breakpoints` and `css` to format responsive arrays
 *
 * @param props - all of the element's props
 * @returns All original props + a formatted, responsive EmotionJS `css` prop
 *
 * @internal usage only
 *
 */

function parseProps(props: any) {
  if (!props || !props.css) return props
  let next: typeof props & { css: Interpolation<any> } = {}

  for (let key in props) {
    if (key === 'css' || key === 'breakpoints') continue
    next[key] = props[key]
  }

  next.css = formatCSS(props.css, props.breakpoints)
  return next
}

/**
 * JSX that supports responsive arrays and the `breakpoints` prop. Uses normal
 * React.createElement if `breakpoints` and `css` are undefined..
 *
 * @remarks
 * This is just a prop-formatting wrapper for Emotion's jsx export
 *
 * @link https://maker-ui.com/docs/jsx
 *
 */

export const jsx = <P extends {}>(
  type: React.ElementType<P>,
  props: React.Attributes & P,
  ...children: React.ReactNode[]
) =>
  typeof props?.css !== 'undefined' || typeof props?.breakpoints !== 'undefined'
    ? emotionJsx(type, parseProps(props), ...children)
    : createElement(type, props, ...children)
