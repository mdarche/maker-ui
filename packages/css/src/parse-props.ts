import { formatCSS } from './css'
import type { Interpolation } from './types'

/**
 * Scans JSX props for `breakpoints` and `css` to format responsive arrays
 *
 * @param props - all of the element's props
 * @returns All original props + a formatted, responsive EmotionJS `css` prop
 *
 * @internal usage only
 *
 */
export function parseProps(props: any) {
  if (!props || !props.css) return props
  let next: typeof props & { css: Interpolation<any> } = {}

  for (let key in props) {
    if (key === 'css' || key === 'breakpoints') continue
    next[key] = props[key]
  }

  next.css = formatCSS(props.css, props.breakpoints)
  return next
}
