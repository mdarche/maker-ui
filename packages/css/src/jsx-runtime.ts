import { createElement } from 'react'
import { jsx as emotionJsx } from '@emotion/react'
import { parseProps } from './parse-props'
import type { MakerProps, StyleObject } from './types'

declare module 'react' {
  interface Attributes extends MakerProps {}
}

/**
 * Scans the CSS prop and returns true if every key is undefined
 *
 * @param {StyleObject} obj a CSS prop style object
 * @returns {boolean}
 */
function isEmpty(obj: StyleObject) {
  return Object.values(obj as object).every(
    (el) => el === undefined || el === null
  )
}

/**
 * JSX that supports responsive arrays and the `breakpoints` prop. Uses normal
 * React.createElement if `breakpoints` and `css` are undefined..
 *
 * @remarks
 * This is just a prop-formatting wrapper for Emotion's jsx export
 *
 * @link https://maker-ui.com/docs/jsx
 *
 */
export const jsx = <P>(
  type: React.ElementType<P>,
  props: React.Attributes & P,
  ...children: React.ReactNode[]
) => {
  if (props?.css && isEmpty(props.css)) {
    delete props.css
    delete props?.breakpoints
  }

  return props?.css
    ? emotionJsx(type as string, parseProps(props), ...children)
    : createElement(type as string, props, ...children)
}
