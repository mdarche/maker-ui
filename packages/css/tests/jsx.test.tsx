/** @jsx jsx */
import { jsx } from '../src/jsx-runtime'
import { createSerializer } from '@emotion/jest'
import { ThemeProvider } from '@emotion/react'
import renderer from 'react-test-renderer'

expect.addSnapshotSerializer(createSerializer())

const Button = (props: any) => <button {...props} />

/**
 * It's difficult to test for media queries. Snapshot testing isn't ideal
 * but Emotion recommends this approach with their @emotion/jest library.
 *
 * @remarks
 * - This test suite is for testing responsive JSX pragma only.
 * - @maker-ui/layout tests for media query / responsive accuracy with Cypress
 * - @maker-ui/primitives package doesn't test for media queries
 */

describe('Custom jsx function', () => {
  test('jsx renders an Emotion component', () => {
    const json = renderer.create(<Button>Test button</Button>).toJSON()

    expect(json).toMatchSnapshot()
  })

  test('identifies and parses the css prop', () => {
    const json = renderer
      .create(<Button css={{ color: 'hotpink' }}>Test button</Button>)
      .toJSON()

    expect(json).toMatchSnapshot()
  })

  test('identifies and parses the breakpoints prop', () => {
    const json = renderer
      .create(
        <Button
          breakpoints={[680, '90vw']}
          css={{ color: ['hotpink', 'red', 'black'] }}>
          Test button
        </Button>
      )
      .toJSON()

    expect(json).toMatchSnapshot()
  })

  test('uses the ThemeProvider default breakpoints if none are specified', () => {
    const json = renderer
      .create(
        <ThemeProvider theme={{ breakpoints: [480, 620] }}>
          <Button css={{ color: ['hotpink', 'red', 'black'] }}>
            Test button
          </Button>
        </ThemeProvider>
      )
      .toJSON()

    expect(json).toMatchSnapshot()
  })

  test('uses default breakpoints if no theme or breakpoints are specified', () => {
    const json = renderer
      .create(
        <Button css={{ color: ['hotpink', 'red', 'black'] }}>
          Test button
        </Button>
      )
      .toJSON()

    expect(json).toMatchSnapshot()
  })
})
