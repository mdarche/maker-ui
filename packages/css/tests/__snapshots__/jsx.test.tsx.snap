// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom jsx function identifies and parses the breakpoints prop 1`] = `
.emotion-0 {
  color: hotpink;
}

@media screen and (min-width: 680px) {
  .emotion-0 {
    color: red;
  }
}

@media screen and (min-width: 90vw) {
  .emotion-0 {
    color: black;
  }
}

<button
  className="emotion-0"
>
  Test button
</button>
`;

exports[`Custom jsx function identifies and parses the css prop 1`] = `
.emotion-0 {
  color: hotpink;
}

<button
  className="emotion-0"
>
  Test button
</button>
`;

exports[`Custom jsx function identifies and parses the custom css prop 1`] = `
.emotion-0 {
  color: hotpink;
}

<button
  className="emotion-0"
>
  Test button
</button>
`;

exports[`Custom jsx function jsx correctly renders an Emotion component 1`] = `
<button>
  Test button
</button>
`;

exports[`Custom jsx function jsx renders an Emotion component 1`] = `
<button>
  Test button
</button>
`;

exports[`Custom jsx function uses default breakpoints if no theme or breakpoints are specified 1`] = `
.emotion-0 {
  color: hotpink;
}

@media screen and (min-width: 768px) {
  .emotion-0 {
    color: red;
  }
}

@media screen and (min-width: 960px) {
  .emotion-0 {
    color: black;
  }
}

<button
  className="emotion-0"
>
  Test button
</button>
`;

exports[`Custom jsx function uses package default breakpoints if no theme or breakpoints prop 1`] = `
.emotion-0 {
  color: hotpink;
}

@media screen and (min-width: 768px) {
  .emotion-0 {
    color: red;
  }
}

@media screen and (min-width: 960px) {
  .emotion-0 {
    color: black;
  }
}

<button
  className="emotion-0"
>
  Test button
</button>
`;

exports[`Custom jsx function uses the ThemeProvider default breakpoints if none are specified 1`] = `
.emotion-0 {
  color: hotpink;
}

@media screen and (min-width: 480px) {
  .emotion-0 {
    color: red;
  }
}

@media screen and (min-width: 620px) {
  .emotion-0 {
    color: black;
  }
}

<button
  className="emotion-0"
>
  Test button
</button>
`;
